airbyte-cdk/python/unit_tests/sources/streams/http/test_http_client.py

```diff
@@ -0,0 +1,312 @@
+    assert prepared_request
+    assert isinstance(prepared_request, requests.PreparedRequest)
+
+def test_connection_pool():
+    http_client = HttpClient(name="test", logger=MagicMock(), authenticator=TokenAuthenticator("test-token"))
+    assert http_client._session.adapters["https://"]._pool_connections == 20
+
+def test_valid_basic_send_request(mocker):
+    http_client = test_http_client()
+    prepared_request = http_client._create_prepared_request(http_method="get", url="https://test_base_url.com/v1/endpoint")
+    mocked_response = MagicMock(spec=requests.Response)
+    mocked_response.status_code = 200
+    mocked_response.headers = {}
+
+    mocker.patch.object(http_client, "_create_prepared_request", return_value=prepared_request)
+    mocker.patch.object(requests.Session, "send", return_value=mocked_response)
+
+    returned_request, returned_response = http_client.send_request(http_method="get", url="https://test_base_url.com/v1/endpoint", request_kwargs={})
+
+    assert returned_request == prepared_request
+    assert returned_response == mocked_response
+
+def test_send_raises_airbyte_traced_exception_with_fail_response_action(mocker):
+    http_client = test_http_client()
+    prepared_request = requests.PreparedRequest()
+    response = requests.Response()
+    response.status_code = 400
+
+
+    mocker.patch.object(http_client._session, "send", return_value=response)
+    mocker.patch.object(http_client._error_handler, "interpret_response", return_value=ErrorResolution(ResponseAction.FAIL, FailureType.system_error, "test error message"))
+
+    with pytest.raises(AirbyteTracedException):
+        http_client._send(prepared_request, {})
+
+    assert http_client._session.send.call_count == 1
+
+def test_send_ignores_with_ignore_reponse_action_and_returns_response(mocker):
+    http_client = test_http_client()
+    prepared_request = http_client._create_prepared_request(http_method="get", url="https://test_base_url.com/v1/endpoint")
+    mocked_response = MagicMock(spec=requests.Response)
+    mocked_response.status_code = 300
+    mocked_response.headers = {}
+    http_client._logger.info = MagicMock()
+
+    mocker.patch.object(requests.Session, "send", return_value=mocked_response)
+    mocker.patch.object(http_client._error_handler, "interpret_response", return_value=ErrorResolution(ResponseAction.IGNORE, FailureType.system_error, "test ignore message"))
+
+    returned_response = http_client._send(prepared_request, {})
+
+    http_client._logger.info.assert_called_once()
+    assert returned_response == mocked_response
+
+@pytest.mark.parametrize(
+        "backoff_time_value, exception_type",
+        [
+            (0.1, UserDefinedBackoffException),
+            (None, DefaultBackoffException)
+        ]
+)
+def test_raises_backoff_exception_with_retry_response_action(mocker, backoff_time_value, exception_type):
```

Comment:
Based on a previous comment, I would assume this will not pass if the error is a 408 for example

Reply:
Just to document -- you are correct. Due to overriding the backoff_time I wasn't invoking the backoff_handler that would cause this to fail. I have since corrected the test and also added an additional `http_client_default_backoff_handler` to manage backoff/retries properly per provided ResponseAction.